File: glass/src/main/java/com/damn/anotherglass/glass/host/ui/MapCard.java
package com.damn.anotherglass.glass.host.ui;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.PowerManager;
import android.support.annotation.NonNull;
import android.util.Log;
import android.widget.RemoteViews;

import com.google.android.glass.timeline.LiveCard;
import com.google.android.glass.widget.CardBuilder;
import com.squareup.picasso.MemoryPolicy;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import java.util.Locale;

public class MapCard extends ICardViewProvider implements LocationListener {

    private static final String LIVE_CARD_TAG = "MapCard";
    private static final int sIMAGE_TIMEOUT = 5000;
    private static final int UPDATE_FREQ_MS = sIMAGE_TIMEOUT;

    private final LocationManager locationManager;
    private final Context mContext;

    private static abstract class TimedTarget implements Target {
        public final long timeStamp = System.currentTimeMillis();
    }

    private Bitmap mLastMap;
    private String mLastMapUrl = "";
    private TimedTarget mMapLoading;

    private Location mLocation;

    @SuppressLint("MissingPermission")
    public MapCard(@NonNull LiveCard card, Context context) {
        super(card);
        mContext = context;
        locationManager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, this);
        updateMapStatus();
    }

    @Override
    public void onRemoved() {
        super.onRemoved();
        locationManager.removeUpdates(this);
    }

    @Override
    public void onLocationChanged(Location location) {
        mLocation = location;
        checkMap();
    }

    @Override
    public void onStatusChanged(String provider, int status, Bundle extras) {

    }

    @Override
    public void onProviderEnabled(String provider) {

    }

    @Override
    public void onProviderDisabled(String provider) {

    }

    private void checkMap() {
        if(null != mMapLoading && mMapLoading.timeStamp + sIMAGE_TIMEOUT > System.currentTimeMillis())
            return;

        if(null == mLocation)
            return;

        final String mapUrl = getMapUrl(mLocation);
        if (mLastMapUrl.equals(mapUrl))
            return;

        mMapLoading = new TimedTarget() {
            @Override
            public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
                mLastMapUrl = mapUrl;
                mLastMap = bitmap;
                mMapLoading = null;
                onMapUpdated();
            }

            @Override
            public void onBitmapFailed(Exception e, Drawable errorDrawable) {
                mMapLoading = null;
                Log.e(LIVE_CARD_TAG, "Failed to load map image " + mapUrl, e);
            }

            @Override
            public void onPrepareLoad(Drawable placeHolderDrawable) {
            }
        };
        Picasso.get().load(mapUrl).memoryPolicy(MemoryPolicy.NO_STORE).into(mMapLoading);
    }

    private void updateMapStatus() {
        PowerManager pm = (PowerManager) mContext.getSystemService(Context.POWER_SERVICE);
        if(null != pm && !pm.isScreenOn()) {
            return;
        }
        checkMap();
    }

    private void onMapUpdated() {
        if (mLiveCard == null || !mLiveCard.isPublished())
            return;
        mLiveCard.setViews(buildView());
    }

    private RemoteViews buildView() {
        CardBuilder cardBuilder = new CardBuilder(mContext.getApplicationContext(), CardBuilder.Layout.TEXT);
        if(null != mLastMap)
            cardBuilder.addImage(mLastMap);
        return cardBuilder.getRemoteViews();
    }

    private String getMapUrl(Location location) {
        return String.format(
                Locale.getDefault(),
                "https://static-maps.yandex.ru/1.x/?lang=en_US" +
                        "&size=640,360&z=15&l=map&pt=" +
                        "%f,%f" +
                        ",pm2rdl",
                location.getLongitude(), location.getLatitude());
    }

}

File: glass-ee/src/main/java/com/damn/anotherglass/glass/ee/host/ui/cards/MapCard.kt
package com.damn.anotherglass.glass.ee.host.ui.cards

import android.annotation.SuppressLint
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import coil3.load
import coil3.request.CachePolicy
import coil3.request.lifecycle
import coil3.request.transitionFactory
import coil3.transition.CrossfadeTransition
import com.damn.anotherglass.glass.ee.host.R
import com.damn.anotherglass.glass.ee.host.databinding.LayoutCardMapBinding
import com.damn.anotherglass.glass.ee.host.gpsPermissions
import com.damn.anotherglass.glass.ee.host.utility.hasPermission
import com.damn.anotherglass.glass.ee.host.utility.locationManager
import java.util.Locale


class MapCard : BaseFragment() {

    private lateinit var locationManager: LocationManager
    private var root: LayoutCardMapBinding? = null
    private var lastMapUrl: String? = null

    private val statusReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (LocationManager.PROVIDERS_CHANGED_ACTION == intent.action) {
                updateState()
            }
        }
    }

    private val locationListener: LocationListener = object : LocationListener {
        override fun onLocationChanged(location: Location) = updateMap(location)
        override fun onProviderEnabled(provider: String) = Unit
        override fun onProviderDisabled(provider: String) = Unit
        @Deprecated("Deprecated in Java", ReplaceWith("Unit"))
        override fun onStatusChanged(provider: String?, status: Int, extras: Bundle?) = Unit
    }

    // todo: write "integrated" location provider, that can also work on broadcasts
    // todo: add zoom in/out menu commands

    private enum class State {
        MissingGeoPermissions,
        MissingMockLocationPermissions,
        MockNotRunning,
        Active,
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        locationManager = context.locationManager()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? = LayoutCardMapBinding.inflate(inflater, container, false)
        .also { root = it }
        .root

    override fun onDestroyView() {
        super.onDestroyView()
        root = null
    }

    override fun onSingleTapUp() {
        super.onSingleTapUp()
        val state = getState()
        @Suppress("DEPRECATION")
        when (state) {
            State.MissingGeoPermissions -> requestPermissions(gpsPermissions, PERMISSIONS_REQUEST_LOCATION)
            State.MissingMockLocationPermissions -> startActivity(Intent(Settings.ACTION_APPLICATION_DEVELOPMENT_SETTINGS))
            State.MockNotRunning -> {} // todo: try to notify service to start mock (if service is running)
            State.Active -> {} // todo: show zoom in/out overlay
        }
    }

    @Deprecated("Deprecated in Java")
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) = when (requestCode) {
        PERMISSIONS_REQUEST_LOCATION -> updateState()
        else -> super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    @SuppressLint("UnspecifiedRegisterReceiverFlag")
    override fun onResume() {
        super.onResume()
        updateState()
        IntentFilter().let {
            it.addAction(LocationManager.PROVIDERS_CHANGED_ACTION)
            requireContext().registerReceiver(statusReceiver, it)
        }
        val context = requireContext()
        lastMapUrl = null // we lost image, force reload
        @SuppressLint("MissingPermission")
        if (hasLocationPermissions(context)) {
            locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER)?.let {
                updateMap(it)
            }
        }
    }

    override fun onPause() {
        super.onPause()
        val context = requireContext()
        if (hasLocationPermissions(context)) {
            locationManager.removeUpdates(locationListener)
        }
        context.unregisterReceiver(statusReceiver)
    }

    private fun updateState() {
        val state = getState()
        @SuppressLint("MissingPermission")
        root?.lblGpsStatus?.text = when (state) {
            State.MissingGeoPermissions -> getString(R.string.msg_gps_permissions_not_granted)
            State.MissingMockLocationPermissions -> getString(R.string.msg_gps_provider_not_available)
            State.MockNotRunning -> getString(R.string.msg_gps_mock_not_running)
            State.Active -> {
                // rate and minimal distance are managed by mobile app, so we can use 0, 0
                locationManager.requestLocationUpdates(
                    LocationManager.GPS_PROVIDER,
                    0,
                    0f,
                    locationListener
                )
                getString(R.string.msg_waiting_for_gps_signal)
            }
        }
    }

    private fun updateMap(location: Location) {
        root?.apply {
            @SuppressLint("SetTextI18n")
            lblGpsStatus.text = "${location.latitude}, ${location.longitude}"
            mapView.apply {
                val url = getMapUrl(location)
                if (lastMapUrl == url) return
                load(url) {
                    diskCachePolicy(CachePolicy.DISABLED)
                    lifecycle(getViewLifecycleOwner())
                    transitionFactory(CrossfadeTransition.Factory())
                    listener(
                        onSuccess = { _, _ -> lastMapUrl = url },
                        onError = { _, _ -> Log.e(TAG, "Failed to load $url") })
                }
            }
        }
    }

    private fun getState(): State = requireContext().let {
        when {
            !hasLocationPermissions(it) -> State.MissingGeoPermissions
            // order is important. Maybe device has a real provider, which is good enough (we should check its active though…)
            locationManager.allProviders.contains(LocationManager.GPS_PROVIDER) -> State.Active
            !hasMockPermission(it) -> State.MissingMockLocationPermissions
            else -> State.MockNotRunning
        }
    }


    companion object {

        private const val TAG = "MapCard"
        private const val PERMISSIONS_REQUEST_LOCATION = 1

        // We need at least one
        private fun hasLocationPermissions(context: Context) =
            gpsPermissions.any { context.hasPermission(it) }

        private fun hasMockPermission(context: Context): Boolean {
            return try {
                val packageInfo = context.packageManager.getPackageInfo(
                    context.packageName,
                    PackageManager.GET_PERMISSIONS
                )
                packageInfo.requestedPermissions?.any { it == "android.permission.ACCESS_MOCK_LOCATION" }
                    ?: false
            } catch (e: Exception) {
                Log.e(TAG, "checkForAllowMockLocationsApps failed: " + e.message)
                // todo:try to install and remove mock provider to check if permission is granted
                false
            }
        }

        // todo: copypasted from Explorer app
        private fun getMapUrl(location: Location): String {
            return String.format(
                Locale.getDefault(),
                "https://static-maps.yandex.ru/1.x/?lang=en_US" +
                        "&size=640,360&z=15&l=map&pt=" +
                        "%f,%f" +
                        ",pm2rdl",
                location.longitude, location.latitude
            )
        }

        @JvmStatic
        fun newInstance(): BaseFragment = MapCard()
    }
}
